src/JobManager.cpp

```diff
@@ -0,0 +1,248 @@
+/**
+ * @file JobManager.cpp
+ *
+ * This file contains definitions for the functions and methods defined in Jobs.hpp.
+ */
+
+#include "einsums/Jobs.hpp"
+#include "einsums/jobs/Job.hpp"
+
+#include <atomic>
+#include <cstdio>
+#include <thread>
+
+/**
+ * @def debug(fmt, ...)
+ *
+ * Print debug information, but only if NDEBUG is undefined.
+ */
+#ifndef NDEBUG
+#    define debug(fmt, ...)                                                                                                                \
+        std::fprintf(stderr, fmt __VA_OPT__(, ) __VA_ARGS__);                                                                              \
+        std::fflush(stderr)
+#else
+#    define debug(fmt, args...)
+#endif
+
+using namespace einsums::jobs;
+
+// Putting this here since it is needed here.
+size_t Job::curr_serial = 0;
+std::mutex Job::serialized_mutex = std::mutex();
+
+/**
+ * @var instance
+ *
+ * Single instance to the job manager.
+ */
+static JobManager *mgr_instance = nullptr;
+
+/**
+ * @var added_job_exit_handler
+ *
+ * Whether the atexit handler has been added to stop and delete the job manager when exit is called, or the program returns from main.
+ */
+
+static bool added_job_exit_handler = false;
+
+JobManager::JobManager() : jobs{}, mutex(), _is_running(false), thread(nullptr) {
+    if (!added_job_exit_handler) {
+        std::atexit(JobManager::cleanup);
+        added_job_exit_handler = true;
+    }
+}
+
+JobManager::~JobManager() {
+    debug("Deleting job manager.\n");
+    this->_is_running = false;
+    if (this->thread != nullptr) {
+        this->thread->join();
+        delete this->thread;
+    }
+    this->jobs.clear();
+}
+
+void JobManager::cleanup() {
+    if (mgr_instance != nullptr) {
+        delete mgr_instance;
+    }
+    mgr_instance = nullptr;
+}
+
+void JobManager::manager_loop() {
+    // Infinite loop.
+    JobManager &inst = JobManager::get_singleton();
+    inst._is_running = true;
+    while (inst._is_running) {
+        inst.manager_event(); // Process an event.
+        std::atomic_thread_fence(std::memory_order_acq_rel);
+        std::this_thread::yield(); // Yield for another thread;
+    }
+    return;
+}
+
+/**
+ * Runs a job.
+ *
+ * @param job The job to run.
+ */
+static void run_job(std::weak_ptr<Job> &job) {
+    std::atomic_thread_fence(std::memory_order_acq_rel);
+    job.lock()->run();
+
+    std::atomic_thread_fence(std::memory_order_acq_rel);
+
+    // Job finished.
+    if (ThreadPool::singleton_exists()) {
+        ThreadPool::get_singleton().thread_finished(std::this_thread::get_id());
+    }
+}
+
+#define JOB_THREADS 8
```

Comment:
Doesn't seem like this is used anywhere.

Reply:
Left that in from probably the first iteration. I'll remove it.